### PR TITLE
Logstructure validation

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

File: libraries/DataFlash/DataFlash.cpp [hunk 1]
```diff
@@ -100,6 +105,129 @@ void DataFlash_Class::Init(const struct LogStructure *structures, uint8_t num_ty
...
+                  fieldcount, labelcount);
+            ret = false;
+            validate_structures_failed();
+        }
+
+        // check that the structure is of an appropriate length to take fields
+        const int16_t msg_len = Log_Write_calc_msg_len(logstructure->format);
+        if (msg_len != logstructure->msg_len) {
+            Debug("Calculated message length for (%s) based on format field (%s) does not match structure size (%d != %u)", logstructure->name, logstructure->format, msg_len, logstructure->msg_len);
+            validate_structures_failed();
```
Comment (OXINARF) [inline libraries/DataFlash/DataFlash.cpp]: Missing `ret = false` (or maybe we should remove that since it will always abort if something is wrong)

Comment (peterbarker) [inline libraries/DataFlash/DataFlash.cpp]: Yep, I've made it void now it is clear this function is always going to be fatal if the structures are invalid.

**Comment by peterbarker** [comment]
I've continued using Debug() here, including for the final "fix your crap" message.  It was confusing having that come out on the console rather than in the SITL window :-)

File: libraries/DataFlash/DataFlash.cpp [hunk 2]
```diff
@@ -100,6 +104,123 @@ void DataFlash_Class::Init(const struct LogStructure *structures, uint8_t num_ty
...
+}
+
+#else
+void DataFlash_Class::dump_structures(const struct LogStructure *structures, const uint8_t num_types)
+{
+}
+
+void DataFlash_Class::validate_structures(const struct LogStructure *structures, const uint8_t num_types)
+{
+    return true;
```
Comment (OXINARF) [inline libraries/DataFlash/DataFlash.cpp]: This has void return now :slightly_smiling_face: Also it is missing an empty implementation of *dump_structure_field*

**Comment by peterbarker** [comment]
On Mon, 3 Apr 2017, Francisco Ferreira wrote:

> This has void return now 🙂 Also it is missing an empty implementation of
> dump_structure_field

Failing to compile not-so-good.  Not really fussed by a dangling 
declaration :-)

**Comment by OXINARF** [comment]
Merged, thanks!

**Pull request merged**